File: csharp/test/Ice/servantLocator/ServantLocatorI.cs
// Copyright (c) ZeroC, Inc.

namespace Ice.servantLocator;

public sealed class ServantLocatorI : Ice.ServantLocator
{
    public ServantLocatorI(string category)
    {
        _category = category;
        _deactivated = false;
        _requestId = -1;
    }

    private static void test(bool b) => global::Test.TestHelper.test(b);

    public Ice.Object locate(Ice.Current curr, out object cookie)
    {
        lock (_mutex)
        {
            test(!_deactivated);
        }

        test(curr.id.category == _category || _category.Length == 0);

        if (curr.id.name == "unknown")
        {
            cookie = null;
            return null;
        }

        if (curr.id.name == "invalidReturnValue" || curr.id.name == "invalidReturnType")
        {
            cookie = null;
            return null;
        }

        test(curr.id.name == "locate" || curr.id.name == "finished");
        if (curr.id.name == "locate")
        {
            exception(curr);
        }

        //
        // Ensure locate() is only called once per request.
        //
        test(_requestId == -1);
        _requestId = curr.requestId;

        cookie = new Cookie();

        return new TestI();
    }

    public void finished(Ice.Current curr, Ice.Object servant, object cookie)
    {
        lock (_mutex)
        {
            test(!_deactivated);
        }

        //
        // Ensure finished() is only called once per request.
        //
        test(_requestId == curr.requestId);
        _requestId = -1;

        test(curr.id.category == _category || _category.Length == 0);
        test(curr.id.name == "locate" || curr.id.name == "finished");

        if (curr.id.name == "finished")
        {
            exception(curr);
        }

        var co = (Cookie)cookie;
        test(co.message() == "blahblah");
    }

    public void deactivate(string category)
    {
        lock (_mutex)
        {
            test(!_deactivated);

            _deactivated = true;
        }
    }

    private void exception(Ice.Current current)
    {
        if (current.operation == "ice_ids")
        {
            throw new Test.TestIntfUserException();
        }
        else if (current.operation == "requestFailedException")
        {
            throw new Ice.ObjectNotExistException();
        }
        else if (current.operation == "unknownUserException")
        {
            throw new Ice.UnknownUserException("reason");
        }
        else if (current.operation == "unknownLocalException")
        {
            throw new Ice.UnknownLocalException("reason");
        }
        else if (current.operation == "unknownException")
        {
            throw new Ice.UnknownException("reason");
        }
        else if (current.operation == "userException")
        {
            throw new Test.TestIntfUserException();
        }
        else if (current.operation == "localException")
        {
            throw new Ice.SocketException();
        }
        else if (current.operation == "csException")
        {
            throw new System.Exception("message");
        }
        else if (current.operation == "unknownExceptionWithServantException")
        {
            throw new Ice.UnknownException("reason");
        }
        else if (current.operation == "impossibleException")
        {
            throw new Test.TestIntfUserException(); // Yes, it really is meant to be TestIntfException.
        }
        else if (current.operation == "intfUserException")
        {
            throw new Test.TestImpossibleException(); // Yes, it really is meant to be TestImpossibleException.
        }
        else if (current.operation == "asyncResponse")
        {
            throw new Test.TestImpossibleException();
        }
        else if (current.operation == "asyncException")
        {
            throw new Test.TestImpossibleException();
        }
    }

    private bool _deactivated;
    private readonly string _category;
    private readonly object _mutex = new();
    private int _requestId;
}

File: swift/test/Ice/servantLocator/ServantLocatorI.swift
// Copyright (c) ZeroC, Inc.

import Foundation
import Ice
import Synchronization
import TestCommon

final class ServantLocatorI: Ice.ServantLocator {
    private class Cookie {
        func message() -> String {
            return "blahblah"
        }
    }

    private struct State {
        var deactivated = false
        var requestId: Int32 = -1
    }

    private let _category: String
    private let _helper: TestHelper
    private let _state = Mutex<State>(State())

    init(_ category: String, _ helper: TestHelper) {
        _category = category
        _helper = helper
    }

    deinit {
        _state.withLock {
            precondition($0.deactivated, "ServantLocatorI deactivated must be true before deinit")
        }
    }

    func locate(_ curr: Ice.Current) throws -> (returnValue: Ice.Dispatcher?, cookie: AnyObject?) {
        try _state.withLock {
            try _helper.test(!$0.deactivated)
        }

        try _helper.test(curr.id.category == _category || _category == "")

        if curr.id.name == "unknown" {
            return (nil, nil)
        }

        if curr.id.name == "invalidReturnValue" || curr.id.name == "invalidReturnType" {
            return (nil, nil)
        }

        try _helper.test(curr.id.name == "locate" || curr.id.name == "finished")

        if curr.id.name == "locate" {
            try exception(curr)
        }

        // Ensure locate() is only called once per request.
        try _state.withLock {
            try _helper.test($0.requestId == -1)
            $0.requestId = curr.requestId
        }

        return (TestI(), Cookie())
    }

    func finished(curr: Ice.Current, servant _: Ice.Dispatcher, cookie: AnyObject?) throws {
        try _state.withLock {
            try _helper.test(!$0.deactivated)

            // Ensure finished() is only called once per request.
            try _helper.test($0.requestId == curr.requestId)
            $0.requestId = -1
        }

        try _helper.test(curr.id.category == _category || _category == "")
        try _helper.test(curr.id.name == "locate" || curr.id.name == "finished")

        if curr.id.name == "finished" {
            try exception(curr)
        }

        try _helper.test((cookie as! Cookie).message() == "blahblah")
    }

    func deactivate(_: String) {
        _state.withLock {
            precondition(!$0.deactivated)
            $0.deactivated = true
        }
    }

    func exception(_ current: Ice.Current) throws {
        if current.operation == "ice_ids" {
            throw TestIntfUserException()
        } else if current.operation == "requestFailedException" {
            throw Ice.ObjectNotExistException()
        } else if current.operation == "unknownUserException" {
            throw Ice.UnknownUserException(badTypeId: "::Foo::BarException")
        } else if current.operation == "unknownLocalException" {
            throw Ice.UnknownLocalException("reason")
        } else if current.operation == "unknownException" {
            throw Ice.UnknownException("reason")
        } else if current.operation == "userException" {
            throw TestIntfUserException()
        } else if current.operation == "localException" {
            throw Ice.SocketException("socket error")
        } else if current.operation == "unknownExceptionWithServantException" {
            throw Ice.UnknownException("reason")
        } else if current.operation == "impossibleException" {
            throw TestIntfUserException()  // Yes, it really is meant to be TestIntfException.
        } else if current.operation == "intfUserException" {
            throw TestImpossibleException()  // Yes, it really is meant to be TestImpossibleException.
        }
    }
}
